components/style/rule_tree/mod.rs

```diff
@@ -169,6 +180,92 @@ impl RuleTree {
     pub unsafe fn maybe_gc(&self) {
         self.root.maybe_gc();
     }
+
+    /// Replaces a rule in a given level (if present) for another rule.
+    ///
+    /// Returns the resulting node that represents the new path.
+    ///
+    /// TODO(emilio): Maybe this should be in `StrongRuleNode`?
+    pub fn replace_rule_at_level_if_applicable(&self,
```

Comment:
So, the if_applicable part of the name here deals with the fact that we might be either inserting or removing, instead of strictly replacing?

How about fn update_rule_at_level(...)?